src/validation/rules/FieldsOnCorrectType.js

```diff
@@ -11,14 +11,70 @@
 import type { ValidationContext } from '../index';
 import { GraphQLError } from '../../error';
 import type { Field } from '../../language/ast';
-import type { GraphQLType } from '../../type/definition';
-
+import {
+  isAbstractType,
+  GraphQLAbstractType,
+  GraphQLObjectType,
+} from '../../type/definition';
 
 export function undefinedFieldMessage(
-  fieldName: string,
-  type: GraphQLType
-): string {
-  return `Cannot query field "${fieldName}" on "${type}".`;
+    field: string,
+    type: string,
+    suggestedTypes: Array<string>): string {
+  let message = `Cannot query field "${field}" on type "${type}".`;
+  const MAX_LENGTH = 5;
+  if (suggestedTypes.length !== 0) {
+    let suggestions = suggestedTypes
+      .slice(0, MAX_LENGTH)
+      .map(t => `"${t}"`)
+      .join(', ');
+    if (suggestedTypes.length > MAX_LENGTH) {
+      suggestions += `, and ${suggestedTypes.length - MAX_LENGTH} other types`;
+    }
+    message += ` However, this field exists on ${suggestions}.`;
+    message += ` Perhaps you meant to use an inline fragment?`;
+  }
+  return message;
+}
+
+/**
+ * Return implementations of `type` that include `fieldName` as a valid field.
+ */
+function getImplementationsIncludingField(
+    type: GraphQLAbstractType,
+    fieldName: string): Array<string> {
+  return type.getPossibleTypes()
+    .filter(t => t.getFields()[fieldName] !== undefined)
+    .map(t => t.name)
+    .sort();
+}
+
+/**
+ * Go through all of the implementations of type, and find other interaces
+ * that they implement. If those interfaces include `field` as a valid field,
+ * return them, sorted by how often the implementations include the other
+ * interface.
+ */
+function getSiblingInterfacesIncludingField(
+    type: GraphQLAbstractType,
+    fieldName: string): Array<string> {
+  const implementingObjects = type.getPossibleTypes()
+    .filter(t => t instanceof GraphQLObjectType);
+
+  const suggestedInterfaces = implementingObjects.reduce((acc, t) => {
```

Comment:
if you wanted this to operate on objects instead of strings, you could do:


```js
const implCount = new Map();

implementingObjectTypes.forEach(objType => {
  objType.getInterfaces().forEach(interfaceType => {
    const count = implCount.get(interfaceType);
    if (count) {
      implCount.set(interfaceType, count + 1);
    } else if (interfaceType.getFields()[fieldName]) {
      implCount.set(interfaceType, 0);
    }
  })
})

return [ ...implCount.values() ].sort(
  (a, b) => implCount.get(b) - implCount.get(a)
);